components/Typography/src/UIFont+MaterialScalable.m

```diff
@@ -0,0 +1,121 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "UIFont+MaterialScalable.h"
+
+#import <objc/runtime.h>
+
+#import "MDCTypography.h"
+#import "UIApplication+AppExtensions.h"
+
+static char MDCFontScaleObjectKey;
+
+@implementation UIFont (MaterialTypography)
+
+- (UIFont *)mdc_scaledFontForSizeCategory:(UIContentSizeCategory)sizeCategory {
+  if (!self.mdc_scalingCurve) {
+    return self;
+  }
+
+  NSNumber *fontSizeNumber;
+  if (sizeCategory) {
+    fontSizeNumber = self.mdc_scalingCurve[sizeCategory];
+  }
+
+  // If you have queried the table for a sizeCategory that doesn't exist, we will return the
+  // traits for XXXL.  This handles the case where the values are requested for one of the
+  // accessibility size categories beyond XXXL such as
+  // UIContentSizeCategoryAccessibilityExtraLarge.  Accessbility size categories are only
```

Comment:
This comment + comment hints at some brittleness here; e.g. what if a custom scaling curve is missing default?

If there is no font size defined for a given scaling curve I'd expect the font not to be adjusted at all; is it possible for us to go with a behavior like that and delete this comment + subsequent code?

```
  if (fontSizeNumber == nil) {
    fontSizeNumber = self.mdc_scalingCurve[UIContentSizeCategoryExtraExtraExtraLarge];
  }
```

Reply:
A scaling curve should be complete.  By complete I mean containing values for all sizes from XXS to XXXL.  The accessibility values are optional, which is why there is the fallback behavior of returning XXXL.  This matches Apple's previous behavior of only scaling body past XXXL.  I'd be happy to change the definition of complete to mean containing values for all sizes from XXS - AXXXL.  (Standard Size categories & A11y extensions)  This would remove the need for fallback behavior.
I think it is acceptable to throw an exception if a curve is missing a size category.
Thoughts?

Reply:
Can we enforce this by creating an object with nonnull properties for each of the expected values, rather than a dictionary? I think this would simultaneously make the code less brittle while also allowing us to encourage good practice when using these APIs.

In other words, instead of using  a `NSDictionary<NSString *, NSNumber *>` as the internal storage, turn that into an object that can be used to store well-typed nonnull values.

Reply:
To be (overly?) specific the key in the dictionary is a UIContentSizeCategory, which admittedly is an NSString * :)  But  NSDictionary<UIContentSizeCategory, NSNumber *> has a lot more context than a vanilla NSString * key.

At some point we are going to have to convert from a UIContentSizeCategory to a CGFloat.  If history indicates we are the *only* people who will be making these curves.  I'm confident that an NSDictionary is as simple as possible and will be easily understood by future devs.  We do give up some safety, but I think the balance is right.